RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h

```diff
@@ -0,0 +1,122 @@
+// user include files
+#include <vector>
+
+
+// functions to select the hits to compute the time of a given cluster
+// start with the only hits with timing information
+// average among the hits contained in the chosen time interval
+
+// N.B. time is corrected wrt vtx-calorimeter distance 
+// with straight line and light speed hypothesis
+// for charged tracks or heavy particles (longer track length or beta < 1) 
+// need to correct the offset at analysis level
+
+
+
+//time-interval based on the smallest one containing a minimum fraction of hits
+float highestDensityFraction(std::vector<float>& hitTimes, float fractionToKeep=0.68 /*fraction between 0 and 1*/, float timeWidthBy=0.5){
+
+  std::sort(hitTimes.begin(), hitTimes.end());
+  int totSize = hitTimes.size();
+
+  int num = 0.;
+  float sum = 0.;
+
+  float minTimeDiff = 999.;
+  int startTimeBin = 0;
+  
+  int startBin = 0;
+  int endBin = totSize;
+
+  int totToKeep = int(totSize*fractionToKeep);
+  int maxStart = totSize - totToKeep;
+
+  for(int ij=0; ij<maxStart; ++ij){
+    float localDiff = fabs(hitTimes[ij] - hitTimes[int(ij+totToKeep)]);
+    if(localDiff < minTimeDiff){
+      minTimeDiff = localDiff;
+      startTimeBin = ij;
+    }
+  }
+
+  // further adjust time width around the chosen one based on the hits density
+  // proved to improve the resolution: get as many hits as possible provided they are close in time
+  startBin = startTimeBin;
+  endBin = int(startBin+totSize*fractionToKeep);
+  float HalfTimeDiff = std::abs(hitTimes[startBin] - hitTimes[endBin]) * timeWidthBy;
+
+  for(int ij=0; ij<startBin; ++ij){
+    if(hitTimes[ij] > (hitTimes[startBin] - HalfTimeDiff) ){
+      for(int kl=ij; kl<totSize; ++kl){
+        if(hitTimes[kl] < (hitTimes[endBin] + HalfTimeDiff) ){
+	  sum += hitTimes[kl];
+	  ++num;
+	}
+	else  break;
+      }
+      break;
+    }
+  }
+  
+  if(num == 0) return -99.;  
+  return sum/num;
+}
+
+
+
+//time-interval based on that ~210ps wide and with the highest number of hits
+float fixSizeHighestDensity(std::vector<float>& t, float deltaT=0.210 /*time window in ns*/, float timeWidthBy=0.5){
+
+  float tolerance = 0.05f;
+  std::sort(t.begin(), t.end());
```

Comment:
this function and the previous function both sort a vector - if they end up sorting the same vector (in future usage), would be nice to isolate that step to avoid repeating it